File: scripts/livelock/livelock.pl
#!/usr/bin/env perl

use v5.38;
use threads;
use threads::shared;
use Time::HiRes qw(sleep);

my $lock_a  :shared = 1;
my $lock_b  :shared = 1;
my $lock_c  :shared = 1;
my $counter :shared = 9;

sub try_lock($this) {
    {
        lock($$this);
        return 1 if $$this;
    }
    sleep(rand() / 10);
    return 0;
}

sub worker($name, $first, $second) {
    while ($counter > 0) {
        {
            lock($$first);
            if ($$first && try_lock($second)) {
                if ($counter > 0) {
                    $counter--;
                    say "$name took the turn, counter is $counter now.";
                }
            }
        }
        sleep(rand() / 10);
    }
}

my @threads;
push @threads, threads->create(\&worker, 'Thread 1', \$lock_a, \$lock_b);
push @threads, threads->create(\&worker, 'Thread 2', \$lock_b, \$lock_c);
push @threads, threads->create(\&worker, 'Thread 3', \$lock_c, \$lock_a);
$_->join foreach @threads;

File: scripts/livelock/livelock.py
#!/usr/bin/env python3

import time
import threading
from random import random

lock_a  = threading.Lock()
lock_b  = threading.Lock()
lock_c  = threading.Lock()
counter = 9

def worker(name, first, second):
    global counter
    while counter > 0:
        first.acquire()
        if not second.acquire(blocking=False):
            first.release()
            time.sleep(random()/10)
        else:
            try:
                if counter > 0:
                    counter -= 1
                    print(f"{name} took the turn, counter is {counter} now.")
            finally:
                second.release()
                first.release()

if __name__ == '__main__':
    threads = []
    threads.append(threading.Thread(target=worker, args=('Thread 1', lock_a, lock_b)))
    threads.append(threading.Thread(target=worker, args=('Thread 2', lock_b, lock_c)))
    threads.append(threading.Thread(target=worker, args=('Thread 3', lock_c, lock_a)))
    for thread in threads:
        thread.start()
    for thread in threads:
        thread.join()
